### PR TITLE
Support NavigationPropertyBinding paths ending in cast segments

### LINKED ISSUE #2502: Support nav property bindings terminating in a type cast segment
In OData 4.01, we added support for a [navigation property binding path](https://docs.oasis-open.org/odata/odata-csdl-xml/v4.01/odata-csdl-xml-v4.01.html#sec_NavigationPropertyPathBinding) that ends in a type cast segment:

"OData 4.01 services MAY have a type-cast segment as the last path segment, allowing to bind instances of different sub-types to different targets."

As per the description, this allows the target set of instances to vary based on the type of the instance.  For example, you could have a collection of items, where widgets come from one entity set and dohickeys come from a different entity set.

We don't currently support this syntax in ODL, OData Client, or WebAPI.

This issue is to track lack of support in ODL.

### Assemblies affected

OData Library 7.x

### Reproduce steps

            var csdl
                = "<?xml version=\"1.0\" encoding=\"utf-16\"?>" +
                    "<edmx:Edmx Version=\"4.0\" xmlns:edmx=\"http://docs.oasis-open.org/odata/ns/edmx\">" +
                      "<edmx:DataServices>" +
                        "<Schema Namespace=\"NS\" xmlns=\"http://docs.oasis-open.org/odata/ns/edm\">" +
                          "<EntityType Name=\"EntityA\">" +
                            "<Key><PropertyRef Name=\"ID\" /></Key>" +
                              "<Property Name=\"ID\" Type=\"Edm.Int32\" Nullable=\"false\" />" +
                              "<NavigationProperty Name=\"Nav\" Type=\"NS.EntityX\" Nullable=\"false\" />" +
                            "</EntityType>" +
                          "<EntityType Name=\"EntityX\" Abstract=\"true\">" +
                            "<Key><PropertyRef Name=\"ID\" /></Key>" +
                            "<Property Name=\"ID\" Type=\"Edm.Int32\" Nullable=\"false\" />" +
                          "</EntityType>" +
                          "<EntityType Name=\"EntityX1\" BaseType=\"NS.EntityX\"/>" +
                          "<EntityType Name=\"EntityX2\" BaseType=\"NS.EntityX\"/>" +
                          "<EntityContainer Name=\"Container\">" +
                            "<EntitySet Name=\"SetA\" EntityType=\"NS.EntityA\">" +
                              "<NavigationPropertyBinding Path=\"Nav/NS.EntityX1\" Target=\"SetX1\" />" +
                              "<NavigationPropertyBinding Path=\"Nav/NS.EntityX2\" Target=\"SetX2\" />" +
                            "</EntitySet>" +
                            "<EntitySet Name=\"SetX1\" EntityType=\"NS.EntityX1\" />" +
                            "<EntitySet Name=\"SetX2\" EntityType=\"NS.EntityX2\" />" +
                          "</EntityContainer>" +
                        "</Schema>" +
                      "</edmx:DataServices>" +
                    "</edmx:Edmx>";
            var model = CsdlReader.Parse(XElement.Parse(csdl).CreateReader());
            var setA = model.FindDeclaredNavigationSource("SetA");
            Assert.NotNull(setA);

            var navProp = setA.EntityType().FindProperty("Nav") as IEdmNavigationProperty;
            Assert.NotNull(navProp);
            Assert.Equal(2, setA.NavigationPropertyBindings.Count());

            // NavPropBinding for for EntityX1 should be SetX1
            var X1NavPropBinding = setA.NavigationPropertyBindings.FirstOrDefault(b => b.Path.PathSegments.Last() == "NS.EntityX1");
            Assert.NotNull(X1NavPropBinding);
            var setX1 = model.FindDeclaredNavigationSource("SetX1");
            Assert.Equal(setX1, X1NavPropBinding.Target);
            Assert.Equal(setX1, setA.FindNavigationTarget(navProp, new EdmPathExpression("Nav/NS.EntityX1")));

            // NavPropBinding for for EntityX2 should be SetX2
            var X2NavPropBinding = setA.NavigationPropertyBindings.FirstOrDefault(b => b.Path.PathSegments.Last() == "NS.EntityX2");
            Assert.NotNull(X2NavPropBinding);
            var setX2 = model.FindDeclaredNavigationSource("SetX2");
            Assert.Equal(setX2, X2NavPropBinding.Target);
            Assert.Equal(setX2, setA.FindNavigationTarget(navProp, new EdmPathExpression("Nav/NS.EntityX2")));

            // Navigation Property without 
            Assert.True(setA.FindNavigationTarget(navProp) is EdmUnknownEntitySet);

### Expected result

Test should pass

### Actual result

Test fails

### Additional detail

Logic needs to be added to CsdlSemanticsNavigationSource.ResolveNavigationPropertyPathForBinding to handle the type cast segment following the navigation property.

### PR DESCRIPTION
**Pull request by mikepizzo**
### Issues

*This pull request fixes #2502.*

### Description

OData 4.01 adds the ability to distinguish nav prop bindings to the same navigation property by type by appending a type segment.

This PR adds support to ODL to have different navigation property bindings to the same navigation property that vary by type.

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### Additional work necessary

Consider only writing nav prop bindings that terminate in type segment for 4.01 responses.

### CODE REVIEW

File: test/FunctionalTests/Microsoft.OData.Edm.Tests/Csdl/CsdlReaderJsonTests.cs [hunk 1]
```diff
@@ -244,6 +246,100 @@ public void ValidateNavigationPropertyBindingPathTraversesNoNonContainmentNaviga
...
+            Assert.Empty(errors);
+
+            var setA = model.FindDeclaredNavigationSource("SetA");
+            Assert.NotNull(setA);
+
+            var navProp = setA.EntityType().FindProperty("Nav") as IEdmNavigationProperty;
+            Assert.NotNull(navProp);
+            Assert.Equal(2, setA.NavigationPropertyBindings.Count());
+
+            // NavPropBinding for for EntityX1 should be SetX1
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Edm.Tests/Csdl/CsdlReaderJsonTests.cs:322]: duplicate 'for'

File: test/FunctionalTests/Microsoft.OData.Edm.Tests/Csdl/CsdlReaderTests.cs [hunk 2]
```diff
@@ -187,6 +187,102 @@ var csdl
...
+
+            var navProp = setA.EntityType().FindProperty("Nav") as IEdmNavigationProperty;
+            Assert.NotNull(navProp);
+            Assert.Equal(2, setA.NavigationPropertyBindings.Count());
+
+            // NavPropBinding for for EntityX1 should be SetX1
+            var X1NavPropBinding = setA.NavigationPropertyBindings.FirstOrDefault(b => b.Path.PathSegments.Last() == "NS.EntityX1");
+            Assert.NotNull(X1NavPropBinding);
+            var setX1 = model.FindDeclaredNavigationSource("SetX1");
+            Assert.Equal(setX1, X1NavPropBinding.Target);
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Edm.Tests/Csdl/CsdlReaderTests.cs:268]: shall we use 'Assert.Same' to make sure they are referencing the same instance?

**Review by xuzhg — DISMISSED** [review]

**Review by habbes — DISMISSED** [review]

**Comment by mikepizzo** [comment]
@habbes, @xuzhg  -- addressed a couple comments and updated tests for 4.0 versus 4.01. please re-review. thanks!

File: src/Microsoft.OData.Core/Microsoft.OData.Core.csproj [hunk 3]
```diff
@@ -81,20 +81,6 @@
     </Content>
   </ItemGroup>
 
-  <ItemGroup>
-    <Compile Update="Microsoft.OData.Core.cs">
-      <DesignTime>True</DesignTime>
-      <AutoGen>True</AutoGen>
-      <DependentUpon>Microsoft.OData.Core.tt</DependentUpon>
-    </Compile>
```
Comment (xuzhg) [inline src/Microsoft.OData.Core/Microsoft.OData.Core.csproj:89]: Are they not needed again?

File: src/Microsoft.OData.Edm/Microsoft.OData.Edm.csproj [hunk 4]
```diff
@@ -47,16 +47,6 @@
   <ItemGroup>
     <Compile Include="..\AssemblyInfo\AssemblyKeys.cs" Link="AssemblyKeys.cs" />
     <Compile Include="..\PlatformHelper.cs" Link="PlatformHelper.cs" />
-    <Compile Update="Microsoft.OData.Edm.cs">
-      <DesignTime>True</DesignTime>
-      <AutoGen>True</AutoGen>
-      <DependentUpon>Microsoft.OData.Edm.tt</DependentUpon>
-    </Compile>
-    <Compile Update="Parameterized.Microsoft.OData.Edm.cs">
```
Comment (xuzhg) [inline src/Microsoft.OData.Edm/Microsoft.OData.Edm.csproj:55]: removed accidently?

File: src/Microsoft.OData.Edm/Schema/EdmNavigationSource.cs [hunk 5]
```diff
@@ -97,7 +97,22 @@ public void AddNavigationTarget(IEdmNavigationProperty navigationProperty, IEdmN
                 return;
             }
 
-            if (navigationProperty.Name != bindingPath.PathSegments.Last())
+            // Last segment must be the navigation segment or a type cast segment
+            string lastSegment = null;
+            string penultimateSegment = null;
+            foreach (string segment in bindingPath.PathSegments)
```
Comment (xuzhg) [inline src/Microsoft.OData.Edm/Schema/EdmNavigationSource.cs:103]: Cannot use "LastOrDefault()"? 
Then check the return is null or not?

File: test/FunctionalTests/Microsoft.OData.Edm.Tests/Csdl/CsdlReaderTests.cs [hunk 6]
```diff
@@ -187,6 +187,108 @@ var csdl
...
+            var navProp = (model.FindType("NS.ComplexB") as IEdmStructuredType).FindProperty("ComplexBNav");
+            Assert.Equal(navProp, set1.NavigationPropertyBindings.First().NavigationProperty);
+        }
+
+        [InlineData("4.0")]
+        [InlineData("4.01")]
+        [Theory]
+        public void ValidateNavigationPropertyBindingPathEndingInTypeCast(string odataVersion)
+        {
+            var entitySetWithNavProperties =
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Edm.Tests/Csdl/CsdlReaderTests.cs:226]: maybe name it as " entitySetWithNavBindings"

File: test/FunctionalTests/Microsoft.OData.Edm.Tests/Csdl/CsdlReaderTests.cs [hunk 7]
```diff
@@ -187,6 +187,108 @@ var csdl
...
+                            "<EntitySet Name=\"SetA\" EntityType=\"NS.EntityA\">" +
+                              "<NavigationPropertyBinding Path=\"Nav/NS.EntityX1\" Target=\"SetX1\" />" +
+                              "<NavigationPropertyBinding Path=\"Nav/NS.EntityX2\" Target=\"SetX2\" />" +
+                            "</EntitySet>";
+            var entitySetWithoutNavProperties =
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Edm.Tests/Csdl/CsdlReaderTests.cs:231]: maybe name it as " entitySetWithoutNavBindings"

File: src/Microsoft.OData.Core/Microsoft.OData.Core.csproj [hunk 8]
```diff
@@ -81,20 +81,6 @@
     </Content>
   </ItemGroup>
```
Comment (xuzhg) [inline src/Microsoft.OData.Core/Microsoft.OData.Core.csproj:82]: It seems there's no source code changed in OData.Core project.

**Review by xuzhg — APPROVED** [review]

**Pull request merged**